Xfem mark

donggiang opened the pull request:
## Reason
Crack tip locations in XFEM code are not consistently updated through the object of **CrackFrontDefinition**. Additionally, to allow crack propagation, additional block to define the crack tip location is required in the input file, while the tip location is already handled by the object **MeshCut2DFractureUserObject**.

## Design
Crack tip locations handled by CrackFrontDefinition  will be automatically updated through a block in the input file defining **MeshCut2DFractureUserObject**. Classes that have CrackFrontDefinition as their member need to initialize it by calling initialSetup() and getting CrackFrontDefinition pointer from **_fe_problem**.
## Impact
This improvement allows consistency of crack front points as the tips move to new elements, avoiding convergence issues in inelastic problems with crack tip enriched solutions.

donggiang committed:
move XFEM_MARK from XFEM to SolidMechanics for registration ref #31417

set execution_order_group=1 for geometricCutUserObject.
The order of user objects is as follows:
 (1) udpate crackFrontDefinition
 (2) update J integral
 (3) update meshCutUserObject
 (4) updateXFEMMesh

lynnmunday commented:
@donggiang We need to split this up into two separate PRs if we can.  One for the XFEM mark flag.  And a seperate PR for the enrichment element changes.  I want to see what changes are needed for only the XFEM mark flag.  I thought we could limit the changes to the solid_mechanics and xfem modules. I want to isolate those changes from the changes we need for the enrichment functions.  

I want to know if we can update the xfem crack tip without making changes to transient.C and fixedPointSolve.C. I can help you with this on Monday.

donggiang commented:
> @donggiang We need to split this up into two separate PRs if we can. One for the XFEM mark flag. And a seperate PR for the enrichment element changes. I want to see what changes are needed for only the XFEM mark flag. I thought we could limit the changes to the solid_mechanics and xfem modules. I want to isolate those changes from the changes we need for the enrichment functions.
> 
> I want to know if we can update the xfem crack tip without making changes to transient.C and fixedPointSolve.C. I can help you with this on Monday.

@lynnmunday I separeted the original branch to two. Now, this branch only has changes to the XFEM_mark flag.

lindsayad reviewed (dismissed):
This doesn't feel right to me. I thought the `ExecFlagRegistry` was for the purpose of not having to pollute execution flags known by the framework with flags from modules

lindsayad commented:
@loganharbour ?

lynnmunday commented on framework/src/base/Moose.C:
```diff
@@ -31,6 +31,7 @@ const ExecFlagType EXEC_INITIAL = registerDefaultExecFlag("INITIAL");
 const ExecFlagType EXEC_LINEAR = registerDefaultExecFlag("LINEAR");
 const ExecFlagType EXEC_LINEAR_CONVERGENCE = registerDefaultExecFlag("LINEAR_CONVERGENCE");
 const ExecFlagType EXEC_NONLINEAR = registerDefaultExecFlag("NONLINEAR");
+const ExecFlagType EXEC_XFEM_MARK = registerDefaultExecFlag("XFEM_MARK");
```
revert this

lynnmunday commented on framework/include/base/Moose.h:
```diff
@@ -122,6 +122,7 @@ extern const ExecFlagType EXEC_LINEAR_CONVERGENCE;
 extern const ExecFlagType EXEC_LINEAR;
 extern const ExecFlagType EXEC_NONLINEAR_CONVERGENCE;
 extern const ExecFlagType EXEC_NONLINEAR;
+extern const ExecFlagType EXEC_XFEM_MARK;
```
revert this

lynnmunday commented on modules/xfem/src/base/XFEMAppTypes.C:
```diff
@@ -8,6 +8,4 @@
-#include "ExecFlagRegistry.h"
-
-const ExecFlagType EXEC_XFEM_MARK = registerDefaultExecFlag("XFEM_MARK");
+#include "ExecFlagRegistry.h"
```
mv this file to solid_mechanics/src/base and rename something like:  solid_mechanicsAppTypes.C and h

Only include the execution flag part of this file. 

In the original XFEMAppTypes.h file move 

typedef unsigned int CutSubdomainID 

to XFEMApp.h right after the includes.

lynnmunday reviewed (changes requested):
First run the xfem and solidmechanics tests with the changes I requested below.  If they fail, try to modify the xfem geometric cut user object executioner order flag to -2.

lindsayad commented:
Those tests are passing in CIVET

donggiang commented:
> Those tests are passing in CIVET

@lindsayad I checkouted upstream/devel, run tests in  module XFEM and got the message above.

bwspenc reviewed (changes requested):
This looks right to me as far as the ordering. As you have this now, we've got the following execution order groups set:
CrackFrontDefinition: -1 (already set in DomainIntegralAction in solid_mechanics)
fracture integral VPPs: 0 (default setting, these are created in DomainIntegralAction)
GeometricCutUserObject: 1 (set in those classes' params, added in this PR)

A few things to do:
1) This is failing the precheck, complaining about trailing whitespace in SolidMechanicsAppTypes.C. It doesn't have trailing whitespace, but it's missing a newline on the last line, which might be triggering it. Try re-saving the file or something to fix that. I was able to do that in my editor. You need to fix that to get the tests to run.
2) Your last commit message is wrong: it says that you set the order to -1, but it was 1. Fix that.
3) I would squash these commits into a single commit since the earlier commits in the history touched framework files that weren't necessary.
4) I think we need to do some more work in the DomainIntegralAction where we're setting the execute flags up. The current behavior there depends on the execute flag being defined in xfem, but I think it isn't right for having it in solid_mechanics. Now I'm wondering whether we even needed to move the execute flag definition at all, actually!
5) It sounds like this does break some tests, so we'll need to investigate those.

donggiang committed:
fixing tests which required new flags to be set in some stress
 user objects and the logic was corrected in the 2d mesh cutter object:

../solid_mechanics/src/actions/DomainIntegralAction.C
This needs to execute on timestep_end to have the correct number of crack
tips when the crack grows.This fix makes this and the cutter mesh have
the same number of crack tips.

src/userobjects/MeshCut2DFractureUserObject.C
Fixed the logic for when a stress and K based stress cutter should be used.
Previously it would never execute stress based growth if a kcrit was being
used even if the stress values were surpassing the critical stress.

bwspenc commented:
It looks like we have a test that fails in parallel: test:solid_mechanics_basic.mesh_cutter_3d/mhs

moosebuild commented:
Job [Valgrind](https://civet.inl.gov/job/3247597/) on f50bf72 : invalidated by @lynnmunday

donggiang commented:
> @donggiang Will you change the commit message for [153c8da](https://github.com/idaholab/moose/commit/153c8da8e4c5313dff78da387d792010a23762da) This is wrong: set execution_order_group=-1 for geometricCutUserObject.
> 
> it should be: set execution_order_group=**1** for geometricCutUserObject!
> 
> I saw this and had to find my notes to figure out the ordering again.

updated

bwspenc commented:
To fix that diffing test, add this line to the 'mhs' test definition in moose/modules/xfem/test/tests/solid_mechanics_basic:
partial=false

bwspenc commented:
@lindsayad I'm going to dismiss your review when the tests pass (which they should this time around) so we can merge this. This no longer touches framework code, which should address your concern.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3248757/), step Generate coverage on 56e8ee8 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Solid mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>a4781e</th>
      <th colspan=4>#31417 56e8ee</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.96%</td>
      <td>84.97%</td>
      <td>+0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>28091</td>
      <td>28099</td>
      <td>+8</td>
      <td>11</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>4972</td>
      <td>4969</td>
      <td>-3</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31417/coverage_diff/solid_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31417/coverage/solid_mechanics)

### Xfem

<table>
  <thead>
    <tr>
      <th></th>
      <th>a4781e</th>
      <th colspan=4>#31417 56e8ee</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.53%</td>
      <td>82.53%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7043</td>
      <td>7044</td>
      <td>+1</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1491</td>
      <td>1491</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31417/coverage_diff/xfem)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31417/coverage/xfem)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31417/coverage/xfem)

</details>



This comment will be updated on new commits.

bwspenc commented:
Test failure is unrelated to this PR. Merging.
